README.md
# multirx
Receive Multiple channels in the bandwidth of an SDR converter (RTL-SDR) optionally recording or streaming to an ICECAST server 

Originally designed to allow Emergency Service Volunteers (Fire, EMT, ...) to
monitor radio channels over internet connected devices and review prior calls.

This was my first attempt at doing anything with Python and GNURadio ... so all
standard disclaimers apply. 

I have also shamelessly used the HAM2MON project as inspiration and general
structure. While HAM2MON is a great curses based app for monitoring ALL activity
in a given bandwidth MULTIRX was designed as a server application to stream and 
record specific channels. 

http://www.rtl-sdr.com/ham2mon-a-nbfm-multi-channel-receiver-for-the-rtl-sdr/

https://github.com/madengr/ham2mon

This is still a work in progress. There is LOTS of opportunity for improvement
in handling error cases or improving overall useability. Constructed comments always
appreciated. Special thanks to my orginal BETA (more like ALPHA) testers: KG2F,
K2CO, and K2RI.

DEPENDANCIES (lots I'm forgetting)
- GNURADIO 
- OSMOSDR


CHANGES
20210911 Maintenance Release

Made some changes to make it more maintainable and configurable.

Moved lots of settings that were hard coded to the XML file

Squelch can be set per channel

A separate postscript can be called per channel (i.e. TWO TONE decode or ID decode)



Recorder.py
#!/usr/bin/env python3

from gnuradio import analog
from gnuradio import audio
from gnuradio import blocks
from gnuradio import eng_notation
from gnuradio import filter
from gnuradio import gr
from gnuradio.eng_option import eng_option
import threading
import subprocess
import time
import os
import logging

class Recorder(gr.hier_block2):
  def __init__(self, samp_rate=16E3, name="unnamed", save_dir=None, postscript=None):
    gr.hier_block2.__init__(self, "Recorder", 
      gr.io_signature(1, 1, gr.sizeof_float), gr.io_signature(1, 1, gr.sizeof_float))

    self.timeout = 18000
    self.save_dir = save_dir
    self.postscript = postscript
    self.record_squelch = analog.pwr_squelch_ff(-200, 0.1, 0, True)
    self.blocks_wavfile_sink = blocks.wavfile_sink("/dev/null", 1, samp_rate, 16)
    self.blocks_null_source = blocks.null_source(gr.sizeof_float*1)
    self.connect(self, (self.record_squelch, 0))
    self.connect((self.record_squelch, 0), (self.blocks_wavfile_sink, 0))
    self.connect((self.blocks_null_source, 0), self)
       
    thread = threading.Thread(target=self.timer_thread, 
      args=(name, self.save_dir, self.blocks_wavfile_sink, \
              self.record_squelch, self.postscript, self.timeout))
    thread.daemon = True
    thread.start()

  def timer_thread(self, name, save_dir, wavfile_sink, rec_squelch, postscript, timeout):

    file_open = False
    on_counter = 0
    off_counter = 0
    sql_open = False

    while True:
      if rec_squelch.unmuted():
        off_counter = 0
        if not sql_open:
          sql_open = True
          logging.info("%s open", name)
        if file_open:
          on_counter+=1
        else:
          tstamp = "_" + time.strftime("%Y%m%d-%H%M%S")
          start_time = time.time()
          file_name = save_dir + "/" + name + tstamp + ".wav"
          wavfile_sink.open(file_name)
          file_open = True
          on_counter = 0
          logging.info("Started new file %s ", file_name)
      else:
        if sql_open:
          sql_open = False
          logging.info("%s closed", name)
        if file_open:
          #close file if idle for 1min or open for <timeout> tenth-seconds
          if off_counter > 600 or on_counter > timeout:
            end_time = time.time()-(off_counter/10)
            if on_counter < timeout:
               logging.info("closing %s after %d seconds of inactivity", \
                      file_name, int(off_counter/10))
               logging.info("Total Activity %d seconds", \
                       (int(on_counter/10)))
            else:
               logging.info("closing %s after %d seconds of activity", \
                      file_name, int(on_counter/10))
            wavfile_sink.close()
            file_open = False
            #delete file if less then 4 seconds
            if on_counter < 40:
              logging.info("deleting %s with only %d seconds activity", \
                  file_name, int(on_counter/10))
              os.unlink(file_name)
            else:
              self.convert_file(file_name, save_dir, name, \
                      postscript, start_time, end_time)
          else:
            off_counter+=1

      time.sleep(0.1)

  def convert_file(self, old_filename, sdir, cname, postscript, stime, etime):
    
    os.putenv("TZ", "EST5EDT")
    new_file_name = sdir + "/" + cname + "/"
    new_file_name += time.strftime("%Y%m%d", time.localtime(stime))
    
    try:
      os.makedirs(new_file_name)
    except OSError:
      pass
    
    new_file_name += "/" + time.strftime("%H%M%S", time.localtime(stime))
    new_file_name += "_"
    new_file_name += time.strftime("%H%M%S", time.localtime(etime))
    new_file_name += ".mp3"
    subprocess.call(['lame', '--silent', old_filename, new_file_name])
    if postscript is not None:
       subprocess.call([postscript, old_filename, new_file_name])
    else:
       os.unlink(old_filename)

class RecTest(gr.top_block):
  def __init__(self):
    gr.top_block.__init__(self, "RecTest")

    samp_rate = 16000
  
    self.source = analog.sig_source_f(samp_rate, analog.GR_SIN_WAVE, 1000, 1, 0)
    self.recorder = Recorder(samp_rate, "testname")
    self.throttle = blocks.throttle(gr.sizeof_float*1, samp_rate,True)
    self.mute = blocks.mute_ff(bool(False))
      
    self.connect((self.source, 0), (self.throttle, 0), (self.mute, 0), (self.recorder, 0))

if __name__ == '__main__':
  
  logging.basicConfig(format='%(asctime)s %(message)s', \
      datefmt='%m/%d/%Y %H:%M:%S', level=logging.DEBUG)
  
  tb = RecTest()
  tb.start()

  print("\nstarting test cycle ...\n")
  tb.mute.set_mute(False)
  print("test loop unmute: 10 second transmission")
  time.sleep(10)
    
  print("test loop mute: Quiet for 65 seconds")
  tb.mute.set_mute(True)
  time.sleep(65)
  
  print("test loop unmute: 3 second transmission")
  tb.mute.set_mute(False)
  time.sleep(3)
    
  print("test loop mute: Quiet for 65 seconds")
  tb.mute.set_mute(True)
  time.sleep(65)
  
  print("test loop unmute: 10 minute transmission")
  tb.mute.set_mute(False)
  time.sleep(600)
    
  print("test loop mute: Quiet for 10 seconds")
  tb.mute.set_mute(True)
  time.sleep(10)
  
  print("test loop unmute: 10 minute transmission")
  tb.mute.set_mute(False)
  time.sleep(600)
    
  print("test loop mute: Quiet for 15 seconds")
  tb.mute.set_mute(True)
  time.sleep(15)
  
  print("test loop unmute: 11 minute transmission")
  tb.mute.set_mute(False)
  time.sleep(660)
    
  print("test loop mute: Quiet for 40 seconds")
  tb.mute.set_mute(True)
  time.sleep(40)
  
  print("test loop unmute: 3 minute transmission")
  tb.mute.set_mute(False)
  time.sleep(180)
    
  print("test loop mute: Quiet for 70 seconds")
  tb.mute.set_mute(True)
  time.sleep(70)
   
  print("\ntest cycle complete\n")
  
  try:
    input('Press Enter to quit: ')
  except EOFError:
    pass
  tb.stop()
  tb.wait()


Demodulator.py
#!/usr/bin/env python3

from gnuradio import analog
from gnuradio import audio
from gnuradio import blocks
from gnuradio import eng_notation
from gnuradio import filter
from gnuradio import gr
from gnuradio.eng_option import eng_option
from gnuradio.filter import firdes
from gnuradio.filter import pfb
from optparse import OptionParser
import math
import numpy as np
import osmosdr
import time
import os
import os.path
import string


class Demodulator(gr.hier_block2):
    def __init__(self, samp_rate=2E6, demod_bb_freq=0, squelch=50, tone=0):
       gr.hier_block2.__init__(self, "Demodulator",
          gr.io_signature(1, 1, gr.sizeof_gr_complex),
          gr.io_signature(1, 1, gr.sizeof_short))

       self.initial_decim = initial_decim = 5
       self.samp_ratio = samp_ratio = samp_rate/1E6
       self.final_rate = final_rate = samp_rate/initial_decim**2/int(samp_rate/1E6)
       self.variable_low_pass_filter_taps_0 = firdes.low_pass(1.0, 1, 0.090, 0.010, firdes.WIN_HAMMING, 6.76)
       self.variable_low_pass_filter_taps_1 = firdes.low_pass(1.0, samp_rate/25, 12.5E3, 1E3, firdes.WIN_HAMMING, 6.76)
       self.variable_low_pass_filter_taps_2 = firdes.low_pass(1.0, final_rate, 3500, 500, firdes.WIN_HAMMING, 6.76)
       self.variable_band_pass_filter_taps_3 = firdes.band_pass(1.0, 16E3, 300, 2800, 100, firdes.WIN_HAMMING, 6.76)
       self.squelch_dB = squelch_dB = -squelch
       self.ctcss_freq = tone
       self.ctcss_samp_rate = int(samp_rate/((self.initial_decim**2)*(samp_rate/1E6)))
       self.final_decim = final_decim = int(samp_rate/1E6)

       self.freq_xlating_fir_filter = filter.freq_xlating_fir_filter_ccc(
           initial_decim, (self.variable_low_pass_filter_taps_0), demod_bb_freq, samp_rate)

       self.fir_filter_0 = filter.fir_filter_ccc(initial_decim, (self.variable_low_pass_filter_taps_0))
       self.fir_filter_0.declare_sample_delay(0)

       self.fir_filter_1 = filter.fir_filter_ccc(int(samp_rate/1E6), (self.variable_low_pass_filter_taps_1))
       self.fir_filter_1.declare_sample_delay(0)

       self.analog_pwr_squelch_0 = analog.pwr_squelch_cc(squelch_dB, 0.1, 0, False)

       self.quadrature_demod = analog.quadrature_demod_cf(1)

       self.analog_ctcss_squelch_0 = analog.ctcss_squelch_ff( \
            self.ctcss_samp_rate, self.ctcss_freq, .01, 0, 0, False)
            
       self.fir_filter_2 = filter.fir_filter_fff(initial_decim, (self.variable_low_pass_filter_taps_2))
       self.fir_filter_2.declare_sample_delay(0)

       self.pfb_arb_resampler_xxx_0 = \
           pfb.arb_resampler_fff(16E3/float(final_rate/5), taps=None, flt_size=32)
       #self.pfb_arb_resampler_xxx_0.declare_sample_delay(0)
       
       self.band_pass_filter = filter.fir_filter_fff(1, (self.variable_band_pass_filter_taps_3))

       self.float_to_short = blocks.float_to_short(scale=32768.0)

       self.connect(self, (self.freq_xlating_fir_filter, 0))
       self.connect((self.freq_xlating_fir_filter, 0), (self.fir_filter_0, 0))    
       self.connect((self.fir_filter_0, 0), (self.fir_filter_1, 0))
       
       self.connect((self.fir_filter_1, 0), (self.analog_pwr_squelch_0, 0))    
       self.connect((self.analog_pwr_squelch_0, 0), (self.quadrature_demod, 0))

       if self.ctcss_freq == 0:
          self.connect((self.quadrature_demod, 0), (self.fir_filter_2, 0))
       else:
          self.connect((self.quadrature_demod, 0), (self.analog_ctcss_squelch_0, 0))
          self.connect((self.analog_ctcss_squelch_0, 0), (self.fir_filter_2, 0))
  
       self.connect((self.fir_filter_2, 0), (self.pfb_arb_resampler_xxx_0, 0))
       self.connect((self.pfb_arb_resampler_xxx_0, 0), (self.band_pass_filter, 0))
       self.connect((self.band_pass_filter, 0), (self.float_to_short, 0))
       self.connect((self.float_to_short, 0), self)
        	
class examplerx(gr.top_block):
    def __init__(self):
        gr.top_block.__init__(self, "MULTIRX testing")

        self.center_freq = 162400000
        self.squelch = -100


        self.samp_rate = samp_rate = 1E6
        self.gain_db = gain_db = 30
        
        self.osmosdr_source = osmosdr.source( args="numchan=" + str(1) + " " + "rtl" )
        self.osmosdr_source.set_sample_rate(samp_rate)
        self.osmosdr_source.set_center_freq(self.center_freq, 0)
        self.osmosdr_source.set_freq_corr(69, 0) #move this to option
        self.osmosdr_source.set_dc_offset_mode(0, 0)
        self.osmosdr_source.set_iq_balance_mode(0, 0)
        self.osmosdr_source.set_gain_mode(False, 0)
        self.osmosdr_source.set_gain(gain_db, 0)
        self.osmosdr_source.set_if_gain(20, 0)
        self.osmosdr_source.set_bb_gain(20, 0)
        self.osmosdr_source.set_antenna("", 0)
        self.osmosdr_source.set_bandwidth(samp_rate*0.8, 0)
	
        self.audio_sink = audio.sink(16000, "", True)
        self.blocks_short_to_float = blocks.short_to_float(1, 32768)

        self.demod_bb_freq = 0
        self.ctcss_freq = 0
        self.demod = Demodulator(self.samp_rate)
        self.connect((self.osmosdr_source, 0), (self.demod, 0))
        self.connect((self.demod, 0), (self.blocks_short_to_float, 0))
        self.connect((self.blocks_short_to_float, 0), (self.audio_sink, 0))    

if __name__ == '__main__':
   tb = examplerx()
   tb.start()
   try:
      input('Press Enter to quit: ')
   except EOFError:
      pass
   tb.stop()
   tb.wait()


configuration.py
#!/usr/bin/env python3

import xml.etree.ElementTree as ET
import sys

class Configuration:
   def __init__(self, filename):
      try:
         tree = ET.parse(filename)
      except IOError:
         sys.exit("configuration file not found.")

      root = tree.getroot()

      self._serial = root.find('system/serial')
      self._gain = root.find('system/gain')
      self._freq_corr = root.find('system/freq_corr')
      self._save_dir = root.find('system/save_dir')
      self._ice_host = root.find('system/ice_host')
      self._ice_port = root.find('system/ice_port')
      self._ice_auth = root.find('system/ice_auth')
   
      # Create an array of tuples with the channel info
      self.channel = []
      for i in root.findall('channel'):
         frequency = int(float(i.find('frequency').text)*1E6)
         key = i.find('key').text
         description = i.find('description').text
         ctcss = float(i.find('ctcss').text)
         _postscript = i.find('postscript')
         if _postscript is not None:
             postscript = _postscript.text
         else:
             postscript = None
         _squelch = i.find('squelch')
         if _squelch is not None:
             squelch = int(_squelch.text)
         else:
             squelch = 20

         self.channel.append((key, frequency, description, ctcss, postscript, squelch))

      # Find the highest Channel
      self.high = 0
      for j in self.channel:
         if j[1] > self.high:
            self.high = j[1]
      
      # Find the lowest Channel
      self.low = float("inf")
      for j in self.channel:
         if j[1] < self.low:
            self.low = j[1]

      #Calculate the span from low to high
      self.width = self.upper() - self.lower()

      #Calculate the center of the span
      self.middle = self.lower() + (self.span()/2)

      
   def ice_host(self):
       if self._ice_host is not None:
          return self._ice_host.text
       else:
          return None

   def ice_port(self):
       if self._ice_port is not None:
          return int(self._ice_port.text)
       else:
          return None

   def ice_auth(self):
       if self._ice_auth is not None:
          return self._ice_auth.text
       else:
          return None

   def save_dir(self):
       if self._save_dir is not None:
          return self._save_dir.text
       else:
          return None

   def freq_corr(self):
       if self._freq_corr is not None:
          return int(self._freq_corr.text)
       else:
          return int(0)

   def gain(self):
       if self._gain is not None:
           return int(self._gain.text)
       else:
           return int(0)

   def serial(self):
       return self._serial.text

   def upper(self):
      return self.high

   def lower(self):
      return self.low

   def center(self):
      return self.middle

   def span(self):
      return self.width

if __name__ == '__main__':

   config = Configuration('receiver2.xml')

   high = config.upper()
   low = config.lower()
   center = config.center()
   span = config.span()
   ser_num = config.serial()
   save_dir = config.save_dir()

   print("Device serial#: %s" % ser_num)
   print("Save Directory: %s" % save_dir)
   print("Gain=%d, Freq_Correction=%d" % 
           (config.gain(), config.freq_corr()))
   print("ice_host: %s, ice_port: %d, ice_auth: %s" % (config.ice_host(), 
           config.ice_port(), config.ice_auth()))
   print("high=%d low=%d" % (high, low))
   print("center=%d" % center)
   print("span=%d" % span)
   for i in config.channel:
      print(i)


ttd-post-example.bash
#!/bin/bash
export TZ=EST5EDT

PAGEDIR=/var/www/html/wav/pager/

sleep 2
/home/user/wave2ttd.py $1 | while read A
do
	mkdir -p ${PAGEDIR}${A}
	NEWFILE=${PAGEDIR}${A}/$(date +%Y-%m-%d_)$(basename $1 .wav).mp3
	/usr/bin/lame --silent $1 ${NEWFILE}
	echo "$(date) $A $NEWFILE" >> /home/user/sdr/multirx2/pages.log

	echo ${NEWFILE} >> /home/user/call_list
	tail -20 /home/user/call_list > /home/user/new_call_list
	rm /home/user/call_list
	mv /home/user/new_call_list /home/user/call_list

#	if [[ $A =~ .*FIRE.* ]]; then
#		/home/user/send-audio-email.py -f xyz@example.com -s "$A Alert" -r 9085551212@vzwpix.com  ${NEWFILE}
#	fi
	if [[ $A =~ 21_RESCUE ]]; then
		/home/user/send-audio-email.py -f xyz@example.com -s "$A Alert" -r sammy@example.com  ${NEWFILE}
	fi
	if [[ $A =~ 21_FIRE ]]; then
		/home/user/send-audio-email.py -f xyz@example.com -s "$A Alert" -r freddy@example.com  ${NEWFILE}
	fi
	if [[ $A =~ 22_HAZMAT ]]; then
		/home/user/push_call.bash ${DIR}${FILE} 19085551212
	fi
done

rm $1


multirx_nogui.py
#!/usr/bin/env python3

from gnuradio import analog
from gnuradio import audio
from gnuradio import blocks
from gnuradio import eng_notation
from gnuradio import filter
from gnuradio import gr
from gnuradio.eng_option import eng_option
from gnuradio.filter import firdes
from gnuradio.filter import pfb
from optparse import OptionParser
import sys
import math
import numpy as np
import osmosdr
import time
import os
import os.path
import subprocess
import threading
import select
import fcntl
import socket
import string
from base64 import b64encode
import configuration
import Recorder
import Demodulator
import logging
                
class multirx_example(gr.top_block):
    def __init__(self, filename, options):
        gr.top_block.__init__(self, "MULTIRX2")

        config = configuration.Configuration(filename)
        self.center_freq = config.center()
        self.save_dir = config.save_dir()

        print("Starting Multirx ...")
        print("low=%d center=%d high=%d span=%d" % (config.lower(), 
           config.center(), config.upper(), 
           (config.upper() - config.lower())))

        self.samp_rate = samp_rate = 1E6

        dev_args="numchan=" + str(1) + " rtl=" + config.serial()
        print("Device Args: %s" % dev_args)
        self.osmosdr_source = osmosdr.source( args=dev_args )
        self.osmosdr_source.set_sample_rate(samp_rate)
        self.osmosdr_source.set_center_freq(self.center_freq, 0)
        self.osmosdr_source.set_freq_corr(config.freq_corr(), 0)
        self.osmosdr_source.set_dc_offset_mode(0, 0)
        self.osmosdr_source.set_iq_balance_mode(0, 0)
        if config.gain() == 0:
           self.osmosdr_source.set_gain_mode(True, 0)
        else:
           self.osmosdr_source.set_gain_mode(False, 0) 
           self.osmosdr_source.set_gain(config.gain(), 0)
        self.osmosdr_source.set_if_gain(20, 0)
        self.osmosdr_source.set_bb_gain(20, 0)
        self.osmosdr_source.set_antenna("", 0)
        self.osmosdr_source.set_bandwidth(0, 0)
        
        self.blocks_adder = blocks.add_vss(1)

        for self.i, self.entry in enumerate(config.channel):
            self.demod_bb_freq = self.entry[1] - self.center_freq
            self.ctcss_freq = self.entry[3]
            self.postscript = self.entry[4]
            self.squelch = self.entry[5]
            self.demod = Demodulator.Demodulator(self.samp_rate, \
                self.demod_bb_freq, self.squelch, self.ctcss_freq)
            self.fname = self.setup_upstream_pipe(self.entry[0], self.entry[2], config)
            self.file_sink = blocks.file_sink(gr.sizeof_short*1, self.fname, True)
            self.connect((self.osmosdr_source, 0), (self.demod, 0), (self.file_sink, 0))    
            self.connect((self.demod, 0), (self.blocks_adder, self.i))
            self.rec = Recorder.Recorder(16000, self.entry[0], self.save_dir, self.postscript)
            self.connect((self.demod, 0), (blocks.short_to_float(1, 32768), 0), (self.rec, 0))
            time.sleep(1) #space things out a bit

        if options.multiplex:        
            self.fname = self.setup_upstream_pipe("multiplex", "Multiplex Channel")
            self.file_sink = blocks.file_sink(gr.sizeof_short*1, self.fname, True)
            self.connect((self.blocks_adder, 0), (self.file_sink, 0))

        if options.local:
            self.audio_sink = audio.sink(16000, "", True)
            self.blocks_short_to_float = blocks.short_to_float(1, 32768)
            self.connect((self.blocks_adder, 0), (self.blocks_short_to_float, 0), (self.audio_sink, 0))
        
        if not options.multiplex and not options.local:
            self.connect((self.blocks_adder, 0), (blocks.null_sink(gr.sizeof_short*1),0))

    def setup_upstream_pipe(self, key, description, config):
        fname = "pipe-%s.raw" % key

        # create a pipe
        try:
            os.unlink(fname)
        except OSError:
            pass

        os.mkfifo(fname)

        samplerate = 16
        bitrate = 16
        bitwidth = 16
        cmd = ['lame',
           '--bitwidth', str(bitwidth),
           '-b', str(bitrate),
           '-m', 'm', # mode mono
           '-r', # raw samples
           '-s', str(samplerate), # sample rate
           '--signed', '--little-endian',
           '--flush', # flush as soon as possible
           '--silent', # less verbose
           '--scale', '1.0',
           fname, '-']

        # maximum bufsize - lame is configured to flush quicker, anyway
        bufsize = 2048
        pipe = subprocess.Popen(cmd, bufsize=bufsize, stdout=subprocess.PIPE).stdout
        # make pipe non-blocking
        fl = fcntl.fcntl(pipe, fcntl.F_GETFL)
        fcntl.fcntl(pipe, fcntl.F_SETFL, fl | os.O_NONBLOCK)

        # set up a thread to read from the pipe
        thread = threading.Thread(target=self.upstream_thread, 
            args=(key, description, pipe, samplerate, bitrate, config))
        thread.daemon = True
        thread.start()

        return fname

    def upstream_thread(self, key, description, pipe, samplerate, bitrate, config):
        """
        Read mpeg data stream from pipe, upload to server with low latency
        """
        global error_count
        # TODO: error handling, never fail - retry and rewire
        poll = select.poll()
        poll.register(pipe, select.POLLIN|select.POLLERR)
        
        ice = None
        last_connect = 0
        
        while True:
            r = poll.poll(1000)
            if len(r) < 1:
                logging.info("%s ... poll timeout", key)
                error_count+=1
                continue
                
            fd, ev = r[0]
            d = pipe.read(4096)
            #print "read %d" % len(d)
            
            if ice == None and time.time() - last_connect > 4:
                # Connect to icecast
                logging.info("%s ... connecting ...", key)
                last_connect = time.time()
                try:
                    ice = self.icecast_connect(key, description, samplerate, bitrate, config)
                    logging.info("%s ... connected!", key)
                except Exception as e:
                    logging.info("... connect failed: %r", e)
                    ice = None
            
            if ice != None:
                try:
                    ice.send(d)
                except Exception:
                    try:
                        ice.close()
                        logging.info("%s ... disconnected!", key)
                    except Exception:
                        pass
                    ice = None

            
    def icecast_connect(self, key, description, samplerate, bitrate, config):
        """
        Connect to icecast
        """
        #key = key + "-test"
        mountpoint = "/%s.mp3" % key

        def request_format(request, line_separator="\n"):
            return line_separator.join(["%s: %s" % (key, str(val)) for (key, val) in request.items()])
        
        host = config.ice_host()
        port = config.ice_port()
        auth = config.ice_auth().encode('ASCII')
            
        logging.info("%s connecting %s:%d", key, host, port)
        
        s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        s.connect((host, port))
        ice_header="SOURCE %s ICE/1.0\n%s\n%s\n\n" % (
            mountpoint,
            request_format({
                'content-type': 'audio/mpeg',
                'Authorization': 'Basic ' + b64encode(b'source:' + auth).decode('ASCII'),
                'User-Agent': "multirx"
            }),
            request_format({
                'ice-name': key,
                'ice-genre': 'Radio',
                'ice-bitrate': bitrate,
                'ice-private': 0,
                'ice-public': 0,
                'ice-description': description,
                'ice-audio-info': "ice-samplerate=%d;ice-bitrate=%d;ice-channels=1" %
                    (samplerate, bitrate),
                'ice-url': "http://%s:%d/%s.mp3" % (host, port, key)
            })
        )
        s.sendall(ice_header.encode('ASCII'))
        response = s.recv(4096)
        if len(response) == 0:
            raise "No response from icecast server"
        
        if response.find(b'HTTP/1.0 200 OK') == -1:
            raise b'Server response: %s' % response
        
        return s
        
if __name__ == '__main__':
   parser=OptionParser(option_class=eng_option)
   parser.add_option("-v", "--verbose", action="store_true", help="verbose")
   parser.add_option("-l", "--local", action="store_true", help="local monitor")
   parser.add_option("-m", "--multiplex", action="store_true", help="multiplex stream")
   (options, args) = parser.parse_args()

   if len(args) != 1:
      parser.print_help()
      sys.exit(1)
   
   logging.basicConfig(format='%(asctime)s %(message)s', \
      datefmt='%m/%d/%Y %H:%M:%S', level=logging.DEBUG)
   error_count = 0
   tb = multirx_example(args[0], options)
   tb.start()
   
   timeout = 5
   print("Press Enter to quit: ")
   rlist, _, _ = select.select([sys.stdin], [], [], timeout)
   while not rlist:
     rlist, _, _ = select.select([sys.stdin], [], [], timeout)
     if error_count > 20:
       logging.info("Error count exceeded, restarting receivers.")
       break
      
   print("Exit Requested.")
   tb.stop()
   tb.wait()
